worlds/tloz/Client.py

```diff
@@ -0,0 +1,377 @@
+            items_received_count_value += 1
+            new_items_received_count_low = items_received_count_value % 256
+            new_items_received_count_high = items_received_count_value // 256
+            new_items_received_count_low_write = (Rom.items_obtained_low, [new_items_received_count_low], self.wram)
+            new_items_received_count_high_write = (Rom.items_obtained_high, [new_items_received_count_high], self.wram)
+            write_list: list[tuple[int, list[int], str]] = [new_items_received_count_low_write, new_items_received_count_high_write]
+            await self.write_item(ctx, current_item_name, write_list)
+
+    async def resolve_shop_items(self, ctx: "BizHawkClientContext"):
+        left_shop_slots = await self.read_ram_value_guarded(ctx, Rom.left_shop_slots)
+        middle_shop_slots = await self.read_ram_value_guarded(ctx, Rom.middle_shop_slots)
+        right_shop_slots = await self.read_ram_value_guarded(ctx, Rom.right_shop_slots)
+        if left_shop_slots is None or middle_shop_slots is None or right_shop_slots is None:
+            return
+        for shop_category, shops in Locations.shop_categories.items():
+            bit = Rom.shop_correspondance[shop_category]
+            for shop in shops:
+                shop_id = Locations.location_table[shop] + base_id
+                if shop_id in ctx.checked_locations:
+                    if "Left" in shop:
+                        left_shop_slots = left_shop_slots | bit
+                    if "Middle" in shop:
+                        middle_shop_slots = middle_shop_slots | bit
+                    if "Right" in shop:
+                        right_shop_slots = right_shop_slots | bit
+        left_shop_write = (Rom.left_shop_slots, [left_shop_slots], self.wram)
+        middle_shop_write = (Rom.middle_shop_slots, [middle_shop_slots], self.wram)
+        right_shop_write = (Rom.right_shop_slots, [right_shop_slots], self.wram)
+        await self.write(ctx, [left_shop_write, middle_shop_write, right_shop_write])
+
+    async def resolve_triforce_fragments(self, ctx: "BizHawkClientContext"):
+        current_triforce_count = await self.read_ram_value_guarded(ctx, Rom.triforce_count)
+        if current_triforce_count is None:
+            return
+        current_triforce_byte = 0xFF >> (8 - min(current_triforce_count, 8))
+        write_list = [(Rom.triforce_fragments, [current_triforce_byte], self.wram)]
+        await self.write(ctx, write_list)
+
+    async def write_item(self, ctx, item_name, write_list: list[tuple[int, list[int], str]]):
+        item_game_id = Items.item_game_ids[item_name]
+        if item_name == "Bomb":
+            item_game_id = 0x29 # Hack to allow bombs to be shown being lifted.
+        lift_write = (Rom.item_to_lift, [item_game_id], self.wram)
+        timer_write = (Rom.item_lift_timer, [128], self.wram) # 128 frames of lifting an item
+        sound_write = (Rom.sound_effect_queue, [4], self.wram) # "Found secret" sound
```

Comment:
And the above comment also reminds me that I think there's no protection here from the client sending items to the player one by one while the player is still in the process of receiving the last one.

Probably this is either accounted for by the game state, or it doesn't matter if you just replace this data mid-animation. But I figured it'd be good to get confirmation just in case.

Reply:
The aesthetic portion of the items being received (lifting the item and playing it) doesn't care if things are rewritten partway through, correct. 

I actually used to have a check for if the item lift timer was zero but forgot it in the initial rewrite...turns out, everyone liked it better because it meant they received bulk item drops faster. Promoted bugs, gotta love them.